### PR TITLE
Ifpack2 : add Zoltan2Partition

### PR DESCRIPTION
**Pull request by iyamazaki**
@trilinos/ifpack2

## Motivation
 This PR adds Zoltan2Partition (e.g., to partition local matrix into blocks for Block Relaxation)

As our first step to having multiple subdomains per MPI for FROSch, we wanted to call Ifpack2’s Block Relaxation as our local subdomain solve. This worked out of the box, calling Tacho on each block. Now, I am hoping to call Zoltan2 to partition the local matrix into blocks.

### CODE REVIEW

File: packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  // Create an input adapter for the Tpetra matrix.
+  Zoltan2GraphAdapterType zoltan2_graph(this->Graph_);
+
+  // Specify partitioning parameters
+  Teuchos::ParameterList zoltan2_params;
+  zoltan2_params.set("partitioning_approach", "partition");
+  zoltan2_params.set("num_local_parts", this->NumLocalParts_);
+  printf( " zoltan2AlgoName = %s\n",zoltan2AlgoName_.c_str() );
+  if (zoltan2AlgoName_ == "parmetis") {
+    printf( " PARMETIS..\n" );
```
Comment (cgcgcg) [inline packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp:85]: remove

**Review by cgcgcg — COMMENTED** [review]
How does this guarantee that partitions are rank-local?

**Comment by iyamazaki** [comment]
@cgcgcg. I thought partitioner computes the partitioning of the Graph passed in, and for Block Relaxation, I thought that is the local Graph for Block Relaxation (which is the case from FROSch since it is local solve). Let me create an ifpack2 tester (e.g., additive Schwarz)..

File: packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    zoltan2_params.set("symmetrize_input", "transpose"); // not sure if this does anything, and may fail with non-symm graph
+    zoltan2_params.set("partitioning_objective", "minimize_cut_edge_weight");
+  } else {
+    zoltan2_params.set("algorithm", zoltan2AlgoName_);
+  }
+
+  // Create and solve partitioning problem
+  Zoltan2::PartitioningProblem<Zoltan2GraphAdapterType>
+    problem(&zoltan2_graph, &zoltan2_params, this->Graph_->getComm());
+  problem.solve();
```
Comment (cgcgcg) [inline packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp:94]: The adapter takes a `Tpetra::CrsGraph` and a communicator. So I'd assume that this is solving an MPI global partitioning problem.

File: packages/ifpack2/src/Ifpack2_Parameters.cpp [hunk 3]
```diff
@@ -170,6 +170,7 @@ void getValidParameters(Teuchos::ParameterList& params)
   // params.get("relaxation: damping factor", 1.0); // already set
   // params.get("relaxation: zero starting solution", true); // already set
   params.set("partitioner: type", "greedy");
+  params.set("zoltan2: algorithm", "phg");
```
Comment (jhux2) [inline packages/ifpack2/src/Ifpack2_Parameters.cpp:173]: Is `phg` a reasonable default?  (Is it a hypergraph partitioner?)

**Comment by jhux2** [comment]
> Let me create an ifpack2 tester (e.g., additive Schwarz).

Yes, it would be helpful to see how you'd like to call this.

**Comment by iyamazaki** [comment]
The way I like to call is as the local solver from FROSch (so just a serial one MPI), but I'll test Block Relaxation as local solve for Ifpack2 additive Schwarz, which may expose the issue.

**Review by jhux2 — CHANGES_REQUESTED** [review]
@iyamazaki Could you add the test you mentioned to this PR?

File: packages/ifpack2/src/Ifpack2_Parameters.cpp (hunk 3 above)
Comment (egboman) [reply packages/ifpack2/src/Ifpack2_Parameters.cpp:173]: phg is a hypergraph partitioner. I think it's reasonable. Metis would be faster but it's a TPL. Jet is a new partitioner that might be better but not in Zoltan2 yet (coming soon).

File: packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp (hunk 2 above)
Comment (egboman) [reply packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp:94]: I think you are right. Even if you set the num_local_parts parameter, I believe Zoltan2 will solve the global partitioning problem, then assign parts to processes. In this use case, it is better to call Zoltan2 with the local graph on each process.

Comment (iyamazaki) [reply packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp:94]: I am trying to figure out a good way to get the local graph. `Ifpack2`'s `LocalFilter` seems to return the graph of the original matrix, and `Tpetra`'s `RowGraph` does not seem to have `getLocalGraph`? @jhux2

Comment (jhux2) [reply packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp:94]: In fact, `LocalFilter` indicates that `getGraph` doesn't do what doxygen says it does.
https://github.com/trilinos/Trilinos/blob/7c431d00fa6873e55b604b8c5fda358f3f050f21/packages/ifpack2/src/Ifpack2_LocalFilter_def.hpp#L288-L298

Comment (jhux2) [reply packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp:94]: In RILUK, the LocalFilter is converted to an actual localized CrsMatrix.   What type of matrix is coming in to this class?

Comment (iyamazaki) [reply packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp:94]: Thank you, @jhux2. I like to get the local part of `A_`, which is of type RowMatrix, before calling Zoltan2 from Block Relaxation, [here](https://github.com/iyamazaki/Trilinos/blob/cd99d2f07eb354e6f1a2656df88c7a95125c8a1d/packages/ifpack2/src/Ifpack2_BlockRelaxation_def.hpp#L658-L659). So, I should follow the same step, and then get the local graph?

**Comment by iyamazaki** [comment]
@jhux2. You mean like GPU streams? I could look for a specific case (e.g., Tacho, though not sure if it will help), but may need your suggestions how to integrate it in a generic way. For now, I am looking performance/numeric without streams. I expect this will still help our setup cost (e.g., storage and computation).

File: packages/ifpack2/src/Ifpack2_BlockRelaxation_def.hpp [hunk 4]
```diff
@@ -650,6 +655,50 @@ initialize ()
...
+          // copy entries into 
+          typename crs_matrix_type::nonconst_local_inds_host_view_type indices("indices",A_local->getLocalMaxNumRowEntries());
+          typename crs_matrix_type::nonconst_values_host_view_type values("values",A_local->getLocalMaxNumRowEntries());
+          for(local_ordinal_type i = 0; i < numRows; i++) {
+            size_t numEntries = 0;
+            A_local->getLocalRowCopy(i, indices, values, numEntries); // get indices & values
+            A_local_graph_nc->insertLocalIndices (i, numEntries, indices.data());
+          }
+          A_local_graph_nc->fillComplete (A_local->getDomainMap (), A_local->getRangeMap ());
+          auto A_local_graph = Teuchos::rcp_const_cast<const crs_graph_type> (A_local_graph_nc);
```
Comment (iyamazaki) [inline packages/ifpack2/src/Ifpack2_BlockRelaxation_def.hpp:692]: I totally misunderstood, but will look if this could run more on GPU. @jhux2

Comment (iyamazaki) [reply packages/ifpack2/src/Ifpack2_BlockRelaxation_def.hpp]: following @jhux2 & @cgcgcg, I moved this code (still on host) into LocalFilter's getGraph function.

**Review by jhux2 — DISMISSED** [review]
This looks good to me, thanks.

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**